{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "\n",
    "import util\n",
    "import train_and_predict\n",
    "import feature_importance\n",
    "import feature_selection\n",
    "from data import *\n",
    "\n",
    "import sklearn\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "path = Path('G:\\My Drive\\Shaked\\WSPC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_genomes_path = path / 'train_genomes.fasta'\n",
    "train_genomes_metedata_path = path / 'train_genomes_info.csv'\n",
    "\n",
    "test_genomes_path = path / 'test_genomes.fasta'\n",
    "test_genomes_metedata_path = path / 'test_genomes_info.csv'\n",
    "\n",
    "train_dataset = GenomesData(train_genomes_path, train_genomes_metedata_path)\n",
    "test_dataset = GenomesData(test_genomes_path, test_genomes_metedata_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check split train dataset to train and validation according to insertion date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train dataset 641\n",
      "train 513 validation 128\n"
     ]
    }
   ],
   "source": [
    "print('train dataset', len(train_dataset.data))\n",
    "\n",
    "split = feature_selection.split_by_insertion_date(train_dataset, proportion=0.2)\n",
    "\n",
    "train_idx, validation_idx = next(split.split())\n",
    "\n",
    "X_train_raw, X_valid_raw = train_dataset.data[train_idx], train_dataset.data[validation_idx]\n",
    "y_train, y_valid = train_dataset.y[train_idx], train_dataset.y[validation_idx]\n",
    "\n",
    "print('train', len(X_train_raw), 'validation', len(X_valid_raw))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    89\n",
       "0    39\n",
       "Name: Label, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_valid.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Timestamp('2018-11-29 03:53:53.656000+0000', tz='UTC')"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATE_INSERTED = 'Date Inserted'\n",
    "\n",
    "min_validation_date = min(train_dataset.metadata.iloc[validation_idx, :][DATE_INSERTED])\n",
    "min_validation_date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Timestamp('2018-11-29 03:08:47.637000+0000', tz='UTC')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_train_date = max(train_dataset.metadata.iloc[train_idx, :][DATE_INSERTED])\n",
    "max_train_date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert min_validation_date > max_train_date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Selection - best k features according to chi2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15% validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best roc_auc score is: 0.9118055555555555\n",
      "{'k_best__k': 600}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_test_balanced_accuracy</th>\n",
       "      <th>mean_test_roc_auc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>param_k_best__k</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.775</td>\n",
       "      <td>0.865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0.811</td>\n",
       "      <td>0.863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>0.819</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>0.811</td>\n",
       "      <td>0.896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>0.806</td>\n",
       "      <td>0.894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0.819</td>\n",
       "      <td>0.895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>0.811</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>0.811</td>\n",
       "      <td>0.899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>0.811</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>0.833</td>\n",
       "      <td>0.903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>0.828</td>\n",
       "      <td>0.903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.842</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mean_test_balanced_accuracy  mean_test_roc_auc\n",
       "param_k_best__k                                                \n",
       "50                                     0.775              0.865\n",
       "100                                    0.811              0.863\n",
       "150                                    0.819              0.897\n",
       "200                                    0.811              0.896\n",
       "250                                    0.806              0.894\n",
       "300                                    0.819              0.895\n",
       "350                                    0.811              0.898\n",
       "400                                    0.811              0.899\n",
       "450                                    0.811              0.901\n",
       "500                                    0.833              0.903\n",
       "550                                    0.828              0.903\n",
       "600                                    0.842              0.912"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val = 50\n",
    "max_val = 600\n",
    "inc = 50\n",
    "\n",
    "k_range = range(min_val, max_val + 1, inc)\n",
    "\n",
    "split = feature_selection.split_by_insertion_date(train_dataset, proportion=0.15)\n",
    "grid_search = feature_selection.perform_fs_k_best(train_dataset, k_range, split=split)\n",
    "print(grid_search.best_params_)\n",
    "\n",
    "param_name = 'k_best__k'\n",
    "feature_selection.grid_search_results_to_df(grid_search, param_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 20% validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best roc_auc score is: 0.9030538749639873\n",
      "{'k_best__k': 450}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_test_balanced_accuracy</th>\n",
       "      <th>mean_test_roc_auc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>param_k_best__k</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.832</td>\n",
       "      <td>0.873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0.828</td>\n",
       "      <td>0.885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>0.813</td>\n",
       "      <td>0.882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>0.815</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>0.826</td>\n",
       "      <td>0.896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0.837</td>\n",
       "      <td>0.896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>0.837</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>0.832</td>\n",
       "      <td>0.899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>0.844</td>\n",
       "      <td>0.903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>0.824</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>0.826</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mean_test_balanced_accuracy  mean_test_roc_auc\n",
       "param_k_best__k                                                \n",
       "50                                     0.832              0.873\n",
       "100                                    0.828              0.885\n",
       "150                                    0.813              0.882\n",
       "200                                    0.815              0.897\n",
       "250                                    0.826              0.896\n",
       "300                                    0.837              0.896\n",
       "350                                    0.837              0.898\n",
       "400                                    0.832              0.899\n",
       "450                                    0.844              0.903\n",
       "500                                    0.824              0.898\n",
       "550                                    0.826              0.897\n",
       "600                                    0.850              0.901"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val = 50\n",
    "max_val = 600\n",
    "inc = 50\n",
    "\n",
    "k_range = range(min_val, max_val + 1, inc)\n",
    "\n",
    "split = feature_selection.split_by_insertion_date(train_dataset, proportion=0.2)\n",
    "grid_search = feature_selection.perform_fs_k_best(train_dataset, k_range, split=split)\n",
    "print(grid_search.best_params_)\n",
    "\n",
    "param_name = 'k_best__k'\n",
    "feature_selection.grid_search_results_to_df(grid_search, param_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5-fold stratified cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best roc_auc score is: 0.9627397654572917\n",
      "{'k_best__k': 550}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_test_balanced_accuracy</th>\n",
       "      <th>mean_test_roc_auc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>param_k_best__k</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.884</td>\n",
       "      <td>0.934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0.889</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>0.894</td>\n",
       "      <td>0.955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>0.900</td>\n",
       "      <td>0.958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>0.904</td>\n",
       "      <td>0.955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0.897</td>\n",
       "      <td>0.956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>0.892</td>\n",
       "      <td>0.958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>0.899</td>\n",
       "      <td>0.957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>0.897</td>\n",
       "      <td>0.961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>0.897</td>\n",
       "      <td>0.962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>0.896</td>\n",
       "      <td>0.963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>0.891</td>\n",
       "      <td>0.961</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mean_test_balanced_accuracy  mean_test_roc_auc\n",
       "param_k_best__k                                                \n",
       "50                                     0.884              0.934\n",
       "100                                    0.889              0.951\n",
       "150                                    0.894              0.955\n",
       "200                                    0.900              0.958\n",
       "250                                    0.904              0.955\n",
       "300                                    0.897              0.956\n",
       "350                                    0.892              0.958\n",
       "400                                    0.899              0.957\n",
       "450                                    0.897              0.961\n",
       "500                                    0.897              0.962\n",
       "550                                    0.896              0.963\n",
       "600                                    0.891              0.961"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val = 50\n",
    "max_val = 600\n",
    "inc = 50\n",
    "\n",
    "k_range = range(min_val, max_val + 1, inc)\n",
    "\n",
    "kf = StratifiedKFold(n_splits=5)\n",
    "grid_search = feature_selection.perform_fs_k_best(train_dataset, k_range, split=kf)\n",
    "print(grid_search.best_params_)\n",
    "\n",
    "param_name = 'k_best__k'\n",
    "feature_selection.grid_search_results_to_df(grid_search, param_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The 5-fold cross validation didn't give consistent results. In my opinion 20% validation set is prefered."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Selection - select best feature from each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "threshold=0.0, selected_features= 450\n",
      "threshold=0.03, selected_features= 406\n",
      "threshold=0.06, selected_features= 373\n",
      "threshold=0.09, selected_features= 329\n",
      "threshold=0.12, selected_features= 298\n",
      "threshold=0.15, selected_features= 270\n",
      "threshold=0.18, selected_features= 244\n",
      "threshold=0.21, selected_features= 219\n",
      "threshold=0.24, selected_features= 201\n",
      "threshold=0.27, selected_features= 181\n",
      "threshold=0.3, selected_features= 159\n",
      "threshold=0.32999999999999996, selected_features= 136\n",
      "threshold=0.36, selected_features= 116\n",
      "threshold=0.39, selected_features= 95\n",
      "threshold=0.42, selected_features= 74\n",
      "threshold=0.44999999999999996, selected_features= 60\n",
      "threshold=0.48, selected_features= 47\n",
      "threshold=0.51, selected_features= 37\n",
      "threshold=0.54, selected_features= 30\n",
      "threshold=0.57, selected_features= 18\n",
      "threshold=0.0, selected_features= 450\n",
      "Best roc_auc score is: 0.9030538749639873\n",
      "{'cluster__threshold': 0.0}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_test_balanced_accuracy</th>\n",
       "      <th>mean_test_roc_auc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>param_cluster__threshold</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0.00</th>\n",
       "      <td>0.844</td>\n",
       "      <td>0.903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.03</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.06</th>\n",
       "      <td>0.843</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.09</th>\n",
       "      <td>0.824</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.12</th>\n",
       "      <td>0.843</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.15</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.18</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.21</th>\n",
       "      <td>0.856</td>\n",
       "      <td>0.898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.24</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.27</th>\n",
       "      <td>0.856</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.30</th>\n",
       "      <td>0.856</td>\n",
       "      <td>0.893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.33</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.36</th>\n",
       "      <td>0.850</td>\n",
       "      <td>0.892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.39</th>\n",
       "      <td>0.856</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.42</th>\n",
       "      <td>0.839</td>\n",
       "      <td>0.897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.45</th>\n",
       "      <td>0.844</td>\n",
       "      <td>0.900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.48</th>\n",
       "      <td>0.844</td>\n",
       "      <td>0.896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.51</th>\n",
       "      <td>0.832</td>\n",
       "      <td>0.876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.54</th>\n",
       "      <td>0.819</td>\n",
       "      <td>0.877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.57</th>\n",
       "      <td>0.824</td>\n",
       "      <td>0.875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          mean_test_balanced_accuracy  mean_test_roc_auc\n",
       "param_cluster__threshold                                                \n",
       "0.00                                            0.844              0.903\n",
       "0.03                                            0.850              0.899\n",
       "0.06                                            0.843              0.898\n",
       "0.09                                            0.824              0.898\n",
       "0.12                                            0.843              0.901\n",
       "0.15                                            0.850              0.901\n",
       "0.18                                            0.850              0.903\n",
       "0.21                                            0.856              0.898\n",
       "0.24                                            0.850              0.900\n",
       "0.27                                            0.856              0.901\n",
       "0.30                                            0.856              0.893\n",
       "0.33                                            0.850              0.892\n",
       "0.36                                            0.850              0.892\n",
       "0.39                                            0.856              0.897\n",
       "0.42                                            0.839              0.897\n",
       "0.45                                            0.844              0.900\n",
       "0.48                                            0.844              0.896\n",
       "0.51                                            0.832              0.876\n",
       "0.54                                            0.819              0.877\n",
       "0.57                                            0.824              0.875"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_val = 0\n",
    "max_val = 0.6\n",
    "inc = 0.03\n",
    "t_range = np.arange(min_val, max_val, inc)\n",
    "\n",
    "# created corr matrix using feature_selection.create_corr_matrix(X_train_raw, y_train) \n",
    "corr_matrix_path = path / 'X_train_corr_mat_0.8_k450.csv'\n",
    "corr_matrix_train = pd.read_csv(corr_matrix_path, index_col=0)\n",
    "dist_matrix_train = feature_selection.feature_corr_to_dist_matrix(corr_matrix_train)\n",
    "\n",
    "split = feature_selection.split_by_insertion_date(train_dataset, proportion=0.2)\n",
    "grid_search_cluster = feature_selection.perform_fs_clusters(train_dataset, dist_matrix_train, t_range, split=split)\n",
    "print(grid_search_cluster.best_params_)\n",
    "\n",
    "param_name = 'cluster__threshold'\n",
    "feature_selection.grid_search_results_to_df(grid_search_cluster, param_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform feature selection process on the entire train dataset, using best parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_k = 450\n",
    "best_t = 0.18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "threshold=0.18, selected_features= 250\n",
      "false_positive: 20,total NHPs: 100\n",
      "false_negative: 7,total HPs: 106\n",
      "BAcc: 0.87\n",
      "sensitivity: 0.93\n",
      "specificity: 0.8\n",
      "aupr_auc: 0.96\n",
      "roc_auc: 0.95\n"
     ]
    }
   ],
   "source": [
    "X_train_raw, X_test_raw = train_dataset.data, test_dataset.data\n",
    "y_train, y_test = train_dataset.y, test_dataset.y\n",
    "\n",
    "# created corr matrix using feature_selection.create_corr_matrix(X_train_raw, y_train) \n",
    "corr_matrix_path = path / 'X_train_corr_mat_k450.csv'\n",
    "X_train_corr_matrix = pd.read_csv(corr_matrix_path, index_col=0)\n",
    "X_train_dist_mat = feature_selection.feature_corr_to_dist_matrix(X_train_corr_matrix)\n",
    "\n",
    "fs_pipeline = Pipeline(steps=[('vectorize', CountVectorizer(lowercase=False, binary=True)),\n",
    "                               ('k_best', SelectKBest(score_func=sklearn.feature_selection.chi2, k=best_k)),\n",
    "                               ('cluster', feature_selection.SelectHierarchicalClustering(X_train_dist_mat, threshold=best_t)),\n",
    "                               ('rf', RandomForestClassifier(random_state=0))])\n",
    "\n",
    "fs_pipeline.fit(X_train_raw, y_train)\n",
    "\n",
    "results = train_and_predict.predict_and_print_results(X_test_raw, y_test, fs_pipeline)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract final features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_pgfams = pd.array(fs_pipeline['vectorize'].get_feature_names())\n",
    "k_best_pgfams = all_pgfams[fs_pipeline['k_best'].get_support()]\n",
    "cluster_pgfams = k_best_pgfams[fs_pipeline['cluster'].get_support()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<641x250 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 41335 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vectorizer = CountVectorizer(lowercase=False, binary=True, vocabulary=cluster_pgfams)\n",
    "vectorizer.fit_transform(X_train_raw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert all(cluster_pgfams == vectorizer.get_feature_names())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final WSPC model\n",
    "A random forest model, which includes the selected genes as features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "false_positive: 20,total NHPs: 100\n",
      "false_negative: 7,total HPs: 106\n",
      "BAcc: 0.87\n",
      "sensitivity: 0.93\n",
      "specificity: 0.8\n",
      "aupr_auc: 0.96\n",
      "roc_auc: 0.95\n"
     ]
    }
   ],
   "source": [
    "WSPC, results = train_and_predict.train_and_predict(X_train_raw, y_train, X_test_raw, y_test, features=cluster_pgfams)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Importance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PGF_00003251</th>\n",
       "      <th>PGF_00004164</th>\n",
       "      <th>PGF_00004169</th>\n",
       "      <th>PGF_00004337</th>\n",
       "      <th>PGF_00006100</th>\n",
       "      <th>PGF_00006245</th>\n",
       "      <th>PGF_00006320</th>\n",
       "      <th>PGF_00006377</th>\n",
       "      <th>PGF_00006380</th>\n",
       "      <th>PGF_00007119</th>\n",
       "      <th>...</th>\n",
       "      <th>PGF_10368922</th>\n",
       "      <th>PGF_10424554</th>\n",
       "      <th>PGF_10470483</th>\n",
       "      <th>PGF_10471787</th>\n",
       "      <th>PGF_10474010</th>\n",
       "      <th>PGF_10506571</th>\n",
       "      <th>PGF_10522720</th>\n",
       "      <th>PGF_10535492</th>\n",
       "      <th>PGF_12758906</th>\n",
       "      <th>PGF_12946886</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>360118.7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90239.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392.338</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1801.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1561.41</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 250 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           PGF_00003251  PGF_00004164  PGF_00004169  PGF_00004337  \\\n",
       "Genome ID                                                           \n",
       "360118.7              0             0             0             0   \n",
       "90239.3               0             0             0             0   \n",
       "1392.338              0             0             0             0   \n",
       "1801.3                0             0             0             0   \n",
       "1561.41               0             0             1             0   \n",
       "\n",
       "           PGF_00006100  PGF_00006245  PGF_00006320  PGF_00006377  \\\n",
       "Genome ID                                                           \n",
       "360118.7              1             0             1             0   \n",
       "90239.3               1             0             1             0   \n",
       "1392.338              0             1             0             0   \n",
       "1801.3                1             0             1             0   \n",
       "1561.41               0             1             0             0   \n",
       "\n",
       "           PGF_00006380  PGF_00007119  ...  PGF_10368922  PGF_10424554  \\\n",
       "Genome ID                              ...                               \n",
       "360118.7              1             0  ...             0             1   \n",
       "90239.3               1             0  ...             0             0   \n",
       "1392.338              0             0  ...             1             0   \n",
       "1801.3                0             0  ...             0             1   \n",
       "1561.41               0             1  ...             1             0   \n",
       "\n",
       "           PGF_10470483  PGF_10471787  PGF_10474010  PGF_10506571  \\\n",
       "Genome ID                                                           \n",
       "360118.7              0             0             1             1   \n",
       "90239.3               0             0             0             0   \n",
       "1392.338              0             0             0             0   \n",
       "1801.3                0             0             1             1   \n",
       "1561.41               0             0             0             0   \n",
       "\n",
       "           PGF_10522720  PGF_10535492  PGF_12758906  PGF_12946886  \n",
       "Genome ID                                                          \n",
       "360118.7              0             1             0             0  \n",
       "90239.3               0             0             0             0  \n",
       "1392.338              0             0             0             0  \n",
       "1801.3                0             1             0             0  \n",
       "1561.41               0             0             0             0  \n",
       "\n",
       "[5 rows x 250 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vectorizer = CountVectorizer(lowercase=False, binary=True, vocabulary=cluster_pgfams)\n",
    "\n",
    "X_train = train_dataset.vectorize_data(vectorizer)\n",
    "y_train = train_dataset.y\n",
    "\n",
    "X_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "pgfams = {}\n",
    "with open(path / 'PATRIC_pgfams.txt') as f:\n",
    "    header = f.readline()\n",
    "    for line in f:\n",
    "        \n",
    "        pgfam = line[:12]\n",
    "        product = line.strip()[13:]\n",
    "        pgfams[pgfam] = product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "class_features = feature_importance.get_top_features_per_class_in_multiple_runs(X_train, y_train, feature_importance.feature_importance, n_runs=100)\n",
    "\n",
    "hps_df, nhps_df = feature_importance.create_top_feats_df(class_features, X_train, y_train, pgfams, top_feats=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Function</th>\n",
       "      <th>Mean Importance (SD)</th>\n",
       "      <th>HPs</th>\n",
       "      <th>NHPs</th>\n",
       "      <th>P-Ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGF_04139053</td>\n",
       "      <td>Uroporphyrinogen III decarboxylase (EC 4.1.1.37)</td>\n",
       "      <td>0.038 (0.012)</td>\n",
       "      <td>362</td>\n",
       "      <td>27</td>\n",
       "      <td>6.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGF_01915472</td>\n",
       "      <td>Dihydrolipoamide acetyltransferase component of pyruvate dehydrogenase complex (EC 2.3.1.12)</td>\n",
       "      <td>0.035 (0.01)</td>\n",
       "      <td>385</td>\n",
       "      <td>48</td>\n",
       "      <td>3.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGF_07629184</td>\n",
       "      <td>Cytosol aminopeptidase PepA (EC 3.4.11.1)</td>\n",
       "      <td>0.03 (0.009)</td>\n",
       "      <td>366</td>\n",
       "      <td>39</td>\n",
       "      <td>4.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGF_07157721</td>\n",
       "      <td>\"Heme O synthase, protoheme IX farnesyltransferase, COX10-CtaB\"</td>\n",
       "      <td>0.022 (0.007)</td>\n",
       "      <td>312</td>\n",
       "      <td>14</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGF_00022550</td>\n",
       "      <td>Molybdopterin synthase catalytic subunit MoaE (EC 2.8.1.12)</td>\n",
       "      <td>0.013 (0.005)</td>\n",
       "      <td>303</td>\n",
       "      <td>17</td>\n",
       "      <td>8.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGF_01033770</td>\n",
       "      <td>Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)</td>\n",
       "      <td>0.011 (0.005)</td>\n",
       "      <td>333</td>\n",
       "      <td>35</td>\n",
       "      <td>4.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PGF_00006100</td>\n",
       "      <td>tRNA-modifying protein YgfZ</td>\n",
       "      <td>0.011 (0.006)</td>\n",
       "      <td>305</td>\n",
       "      <td>17</td>\n",
       "      <td>8.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PGF_07941512</td>\n",
       "      <td>23S rRNA (uracil(1939)-C(5))-methyltransferase (EC 2.1.1.190)</td>\n",
       "      <td>0.01 (0.003)</td>\n",
       "      <td>324</td>\n",
       "      <td>37</td>\n",
       "      <td>4.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PGF_00405499</td>\n",
       "      <td>\"YpfJ protein, zinc metalloprotease superfamily\"</td>\n",
       "      <td>0.009 (0.003)</td>\n",
       "      <td>273</td>\n",
       "      <td>13</td>\n",
       "      <td>10.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PGF_06757295</td>\n",
       "      <td>Threonine dehydratase biosynthetic (EC 4.3.1.19)</td>\n",
       "      <td>0.008 (0.004)</td>\n",
       "      <td>323</td>\n",
       "      <td>34</td>\n",
       "      <td>4.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PGF_08199696</td>\n",
       "      <td>Glutathione reductase (EC 1.8.1.7)</td>\n",
       "      <td>0.008 (0.002)</td>\n",
       "      <td>220</td>\n",
       "      <td>12</td>\n",
       "      <td>9.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PGF_03081665</td>\n",
       "      <td>\"Cell division integral membrane protein, YggT and half-length relatives\"</td>\n",
       "      <td>0.007 (0.003)</td>\n",
       "      <td>352</td>\n",
       "      <td>56</td>\n",
       "      <td>3.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PGF_07854425</td>\n",
       "      <td>Superoxide dismutase [Cu-Zn] precursor (EC 1.15.1.1)</td>\n",
       "      <td>0.006 (0.001)</td>\n",
       "      <td>252</td>\n",
       "      <td>21</td>\n",
       "      <td>5.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PGF_01668012</td>\n",
       "      <td>Sulfur carrier protein FdhD</td>\n",
       "      <td>0.006 (0.003)</td>\n",
       "      <td>300</td>\n",
       "      <td>26</td>\n",
       "      <td>5.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PGF_01147190</td>\n",
       "      <td>Deoxyribodipyrimidine photolyase (EC 4.1.99.3)</td>\n",
       "      <td>0.006 (0.003)</td>\n",
       "      <td>281</td>\n",
       "      <td>17</td>\n",
       "      <td>8.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Feature  \\\n",
       "0   PGF_04139053   \n",
       "1   PGF_01915472   \n",
       "2   PGF_07629184   \n",
       "3   PGF_07157721   \n",
       "4   PGF_00022550   \n",
       "5   PGF_01033770   \n",
       "6   PGF_00006100   \n",
       "7   PGF_07941512   \n",
       "8   PGF_00405499   \n",
       "9   PGF_06757295   \n",
       "10  PGF_08199696   \n",
       "11  PGF_03081665   \n",
       "12  PGF_07854425   \n",
       "13  PGF_01668012   \n",
       "14  PGF_01147190   \n",
       "\n",
       "                                                                                        Function  \\\n",
       "0                                               Uroporphyrinogen III decarboxylase (EC 4.1.1.37)   \n",
       "1   Dihydrolipoamide acetyltransferase component of pyruvate dehydrogenase complex (EC 2.3.1.12)   \n",
       "2                                                      Cytosol aminopeptidase PepA (EC 3.4.11.1)   \n",
       "3                                \"Heme O synthase, protoheme IX farnesyltransferase, COX10-CtaB\"   \n",
       "4                                    Molybdopterin synthase catalytic subunit MoaE (EC 2.8.1.12)   \n",
       "5                                            Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)   \n",
       "6                                                                    tRNA-modifying protein YgfZ   \n",
       "7                                  23S rRNA (uracil(1939)-C(5))-methyltransferase (EC 2.1.1.190)   \n",
       "8                                               \"YpfJ protein, zinc metalloprotease superfamily\"   \n",
       "9                                               Threonine dehydratase biosynthetic (EC 4.3.1.19)   \n",
       "10                                                            Glutathione reductase (EC 1.8.1.7)   \n",
       "11                     \"Cell division integral membrane protein, YggT and half-length relatives\"   \n",
       "12                                          Superoxide dismutase [Cu-Zn] precursor (EC 1.15.1.1)   \n",
       "13                                                                   Sulfur carrier protein FdhD   \n",
       "14                                                Deoxyribodipyrimidine photolyase (EC 4.1.99.3)   \n",
       "\n",
       "   Mean Importance (SD)  HPs  NHPs  P-Ratio  \n",
       "0         0.038 (0.012)  362    27     6.67  \n",
       "1          0.035 (0.01)  385    48     3.99  \n",
       "2          0.03 (0.009)  366    39     4.67  \n",
       "3         0.022 (0.007)  312    14    11.09  \n",
       "4         0.013 (0.005)  303    17     8.87  \n",
       "5         0.011 (0.005)  333    35     4.73  \n",
       "6         0.011 (0.006)  305    17     8.93  \n",
       "7          0.01 (0.003)  324    37     4.36  \n",
       "8         0.009 (0.003)  273    13    10.45  \n",
       "9         0.008 (0.004)  323    34     4.73  \n",
       "10        0.008 (0.002)  220    12     9.12  \n",
       "11        0.007 (0.003)  352    56     3.13  \n",
       "12        0.006 (0.001)  252    21     5.97  \n",
       "13        0.006 (0.003)  300    26     5.74  \n",
       "14        0.006 (0.003)  281    17     8.23  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hps_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Function</th>\n",
       "      <th>Mean Importance (SD)</th>\n",
       "      <th>HPs</th>\n",
       "      <th>NHPs</th>\n",
       "      <th>P-Ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGF_03029062</td>\n",
       "      <td>\"Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)\"</td>\n",
       "      <td>0.034 (0.01)</td>\n",
       "      <td>70</td>\n",
       "      <td>182</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGF_01667671</td>\n",
       "      <td>Cytidylate kinase (EC 2.7.4.25)</td>\n",
       "      <td>0.031 (0.009)</td>\n",
       "      <td>7</td>\n",
       "      <td>125</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGF_02930287</td>\n",
       "      <td>Reverse rubrerythrin</td>\n",
       "      <td>0.027 (0.01)</td>\n",
       "      <td>11</td>\n",
       "      <td>125</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGF_08946513</td>\n",
       "      <td>Flavodoxin</td>\n",
       "      <td>0.021 (0.007)</td>\n",
       "      <td>67</td>\n",
       "      <td>170</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGF_01469197</td>\n",
       "      <td>\"RNA methyltransferase, TrmA family\"</td>\n",
       "      <td>0.021 (0.009)</td>\n",
       "      <td>76</td>\n",
       "      <td>180</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PGF_01333294</td>\n",
       "      <td>Activator of (R)-2-hydroxyglutaryl-CoA dehydratase</td>\n",
       "      <td>0.019 (0.007)</td>\n",
       "      <td>25</td>\n",
       "      <td>128</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PGF_01284176</td>\n",
       "      <td>Rubrerythrin</td>\n",
       "      <td>0.018 (0.009)</td>\n",
       "      <td>35</td>\n",
       "      <td>142</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PGF_00006245</td>\n",
       "      <td>Formate--tetrahydrofolate ligase (EC 6.3.4.3)</td>\n",
       "      <td>0.016 (0.007)</td>\n",
       "      <td>138</td>\n",
       "      <td>201</td>\n",
       "      <td>0.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PGF_00033940</td>\n",
       "      <td>Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)</td>\n",
       "      <td>0.016 (0.006)</td>\n",
       "      <td>6</td>\n",
       "      <td>106</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PGF_10332317</td>\n",
       "      <td>Electron transport complex protein RnfB</td>\n",
       "      <td>0.014 (0.006)</td>\n",
       "      <td>18</td>\n",
       "      <td>123</td>\n",
       "      <td>0.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PGF_08126536</td>\n",
       "      <td>Flavoprotein</td>\n",
       "      <td>0.013 (0.006)</td>\n",
       "      <td>13</td>\n",
       "      <td>114</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PGF_00401757</td>\n",
       "      <td>no significant homology.</td>\n",
       "      <td>0.013 (0.006)</td>\n",
       "      <td>8</td>\n",
       "      <td>106</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PGF_00016404</td>\n",
       "      <td>\"LSU ribosomal protein L32p @ LSU ribosomal protein L32p, zinc-dependent\"</td>\n",
       "      <td>0.012 (0.005)</td>\n",
       "      <td>88</td>\n",
       "      <td>173</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PGF_00075770</td>\n",
       "      <td>FIG00519347: Ribonucleotide reductase-like protein</td>\n",
       "      <td>0.012 (0.003)</td>\n",
       "      <td>51</td>\n",
       "      <td>135</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PGF_00003251</td>\n",
       "      <td>Macrolide export ATP-binding/permease protein</td>\n",
       "      <td>0.01 (0.002)</td>\n",
       "      <td>16</td>\n",
       "      <td>69</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Feature  \\\n",
       "0   PGF_03029062   \n",
       "1   PGF_01667671   \n",
       "2   PGF_02930287   \n",
       "3   PGF_08946513   \n",
       "4   PGF_01469197   \n",
       "5   PGF_01333294   \n",
       "6   PGF_01284176   \n",
       "7   PGF_00006245   \n",
       "8   PGF_00033940   \n",
       "9   PGF_10332317   \n",
       "10  PGF_08126536   \n",
       "11  PGF_00401757   \n",
       "12  PGF_00016404   \n",
       "13  PGF_00075770   \n",
       "14  PGF_00003251   \n",
       "\n",
       "                                                                     Function  \\\n",
       "0    \"Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)\"   \n",
       "1                                             Cytidylate kinase (EC 2.7.4.25)   \n",
       "2                                                        Reverse rubrerythrin   \n",
       "3                                                                  Flavodoxin   \n",
       "4                                        \"RNA methyltransferase, TrmA family\"   \n",
       "5                          Activator of (R)-2-hydroxyglutaryl-CoA dehydratase   \n",
       "6                                                                Rubrerythrin   \n",
       "7                               Formate--tetrahydrofolate ligase (EC 6.3.4.3)   \n",
       "8      Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)   \n",
       "9                                     Electron transport complex protein RnfB   \n",
       "10                                                               Flavoprotein   \n",
       "11                                                   no significant homology.   \n",
       "12  \"LSU ribosomal protein L32p @ LSU ribosomal protein L32p, zinc-dependent\"   \n",
       "13                         FIG00519347: Ribonucleotide reductase-like protein   \n",
       "14                              Macrolide export ATP-binding/permease protein   \n",
       "\n",
       "   Mean Importance (SD)  HPs  NHPs  P-Ratio  \n",
       "0          0.034 (0.01)   70   182     0.19  \n",
       "1         0.031 (0.009)    7   125     0.03  \n",
       "2          0.027 (0.01)   11   125     0.04  \n",
       "3         0.021 (0.007)   67   170     0.20  \n",
       "4         0.021 (0.009)   76   180     0.21  \n",
       "5         0.019 (0.007)   25   128     0.10  \n",
       "6         0.018 (0.009)   35   142     0.12  \n",
       "7         0.016 (0.007)  138   201     0.34  \n",
       "8         0.016 (0.006)    6   106     0.03  \n",
       "9         0.014 (0.006)   18   123     0.07  \n",
       "10        0.013 (0.006)   13   114     0.06  \n",
       "11        0.013 (0.006)    8   106     0.04  \n",
       "12        0.012 (0.005)   88   173     0.25  \n",
       "13        0.012 (0.003)   51   135     0.19  \n",
       "14         0.01 (0.002)   16    69     0.12  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nhps_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_features = ['PGF_07157721', \n",
    "                'PGF_00022550', \n",
    "                'PGF_00405499',\n",
    "                'PGF_06594013',\n",
    "                'PGF_06757295',\n",
    "                'PGF_07941512',\n",
    "                'PGF_00006461',\n",
    "                'PGF_00037937',\n",
    "                'PGF_03081665',\n",
    "                'PGF_02905791',\n",
    "                'PGF_08199696',\n",
    "                'PGF_01147190',\n",
    "                'PGF_07854425',\n",
    "                'PGF_01761390',\n",
    "                'PGF_09847065'\n",
    "               ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(hps_df['Feature'][:15]).intersection(old_features))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test according to groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "GROUP = 'Group'\n",
    "GENOME_ID = 'Genome ID'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Label</th>\n",
       "      <th>HP/NHP entire dataset</th>\n",
       "      <th>species</th>\n",
       "      <th>References</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>163603.4</th>\n",
       "      <td>163603.4</td>\n",
       "      <td>Actinomadura latina strain ATCC BAA-277</td>\n",
       "      <td>HP</td>\n",
       "      <td>1/0</td>\n",
       "      <td>Actinomadura latina</td>\n",
       "      <td>\\cite{trujillo1997polyphasic}</td>\n",
       "      <td>HP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648.157</th>\n",
       "      <td>648.157</td>\n",
       "      <td>Aeromonas caviae strain ScAc2001</td>\n",
       "      <td>HP</td>\n",
       "      <td>9/0</td>\n",
       "      <td>Aeromonas caviae</td>\n",
       "      <td>\\cite{tang2020co}</td>\n",
       "      <td>HP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565.15</th>\n",
       "      <td>565.15</td>\n",
       "      <td>Atlantibacter hermannii strain 3608</td>\n",
       "      <td>HP</td>\n",
       "      <td>1/0</td>\n",
       "      <td>Atlantibacter hermannii</td>\n",
       "      <td>\\cite{ioannou2019escherichia}</td>\n",
       "      <td>HP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2026190.21</th>\n",
       "      <td>2026190.21</td>\n",
       "      <td>Bacillus mobilis strain 1428.155</td>\n",
       "      <td>HP</td>\n",
       "      <td>1/0</td>\n",
       "      <td>Bacillus mobilis</td>\n",
       "      <td>\\cite{2026190.21}</td>\n",
       "      <td>HP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29459.655</th>\n",
       "      <td>29459.655</td>\n",
       "      <td>Brucella melitensis strain HN20190002</td>\n",
       "      <td>HP</td>\n",
       "      <td>13/0</td>\n",
       "      <td>Brucella melitensis</td>\n",
       "      <td>\\cite{li2020molecular}</td>\n",
       "      <td>HP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genome ID                              Genome Name Label  \\\n",
       "Genome ID                                                               \n",
       "163603.4      163603.4  Actinomadura latina strain ATCC BAA-277    HP   \n",
       "648.157        648.157         Aeromonas caviae strain ScAc2001    HP   \n",
       "565.15          565.15      Atlantibacter hermannii strain 3608    HP   \n",
       "2026190.21  2026190.21         Bacillus mobilis strain 1428.155    HP   \n",
       "29459.655    29459.655    Brucella melitensis strain HN20190002    HP   \n",
       "\n",
       "           HP/NHP entire dataset                  species  \\\n",
       "Genome ID                                                   \n",
       "163603.4                     1/0      Actinomadura latina   \n",
       "648.157                      9/0         Aeromonas caviae   \n",
       "565.15                       1/0  Atlantibacter hermannii   \n",
       "2026190.21                   1/0         Bacillus mobilis   \n",
       "29459.655                   13/0      Brucella melitensis   \n",
       "\n",
       "                               References Group  \n",
       "Genome ID                                        \n",
       "163603.4    \\cite{trujillo1997polyphasic}    HP  \n",
       "648.157                 \\cite{tang2020co}    HP  \n",
       "565.15      \\cite{ioannou2019escherichia}    HP  \n",
       "2026190.21              \\cite{2026190.21}    HP  \n",
       "29459.655          \\cite{li2020molecular}    HP  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_genome_verified = pd.read_csv('test_genomes_verified.csv', dtype=str, index_col=0)\n",
    "test_genome_verified.index = test_genome_verified[GENOME_ID]\n",
    "test_genome_verified.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HP-OPP     76\n",
       "NHP        61\n",
       "NHP-OPP    41\n",
       "HP         26\n",
       "Name: Group, dtype: int64"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_genome_verified[GROUP].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HP\taccuracy=1.00 correctly predicted=26/26 average score=0.94\n",
      "HP-OPP\taccuracy=0.92 correctly predicted=70/76 average score=0.90\n",
      "NHP\taccuracy=0.92 correctly predicted=56/61 average score=0.13\n",
      "NHP-OPP\taccuracy=0.61 correctly predicted=25/41 average score=0.43\n"
     ]
    }
   ],
   "source": [
    "for group_name, genomes in test_genome_verified.groupby(GROUP):\n",
    "    \n",
    "    X_test_group = X_test_raw[genomes.index]\n",
    "    y_test_group = y_test[genomes.index]\n",
    "    \n",
    "    y_pred = WSPC.predict(X_test_group)\n",
    "    correct = sklearn.metrics.accuracy_score(y_test_group, y_pred, normalize=False)\n",
    "    accuracy = sklearn.metrics.accuracy_score(y_test_group, y_pred)\n",
    "    \n",
    "    y_probs = WSPC.predict_proba(X_test_group)[:, 1]\n",
    "    average_score = np.average(y_probs)\n",
    "    print(f'{group_name}\\taccuracy={accuracy:.2f} correctly predicted={correct}/{len(y_test_group)} average score={average_score:.2f}')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
