{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "muslim-discussion",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import astropy\n",
    "from astropy.table import Table\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.metrics import classification_report\n",
    "\n",
    "import sys  \n",
    "sys.path.insert(0, '../../Code')\n",
    "\n",
    "import util\n",
    "from data import *\n",
    "import train_and_predict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "light-roots",
   "metadata": {},
   "source": [
    "# Download Bacpacs Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "historical-highlight",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ftplib import FTP\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "FILE_SUFFIX = '.PATRIC.features.tab'\n",
    "logger = open('logger', 'w')\n",
    "\n",
    "def read_genome_ids(input_file):\n",
    "    with open(input_file) as f:\n",
    "        ids = f.read().strip().split('\\n')\n",
    "    return ids\n",
    "\n",
    "def grab_file(ftp, filename, local_path):\n",
    "\n",
    "    local_file = open(local_path, 'wb')\n",
    "\n",
    "    try:\n",
    "        ftp.retrbinary('RETR ' + filename, local_file.write, 1024)\n",
    "    except Exception as e:\n",
    "        logger.write(e)\n",
    "\n",
    "    local_file.close()\n",
    "\n",
    "\n",
    "def download_genomes(genomes, output_dir_path):\n",
    "\n",
    "\n",
    "    # connect to host, default port\n",
    "    ftp = FTP(r'ftp.patricbrc.org')\n",
    "\n",
    "    ftp.login()  # user anonymous, passwd anonymous@\n",
    "\n",
    "    count = 0\n",
    "    for i, genome_id in enumerate(genomes):\n",
    "\n",
    "        try:\n",
    "            ftp.cwd(\"/genomes/\" + genome_id)\n",
    "            local_file = output_dir_path + genome_id + FILE_SUFFIX\n",
    "            remote_file = genome_id + FILE_SUFFIX\n",
    "            grab_file(ftp, remote_file, local_file)\n",
    "\n",
    "            print('downloaded genome: ' + genome_id + ' index: ' + str(i) + '\\n')\n",
    "            count += 1\n",
    "\n",
    "        except Exception as e:\n",
    "            # print('ERROR genome: ' + genome_id + ' index: ' + str(i) + '\\n')\n",
    "            logger.write(f'genome: {genome_id} index: {i} error: {e}\\n')\n",
    "\n",
    "    print('downloaded {} genomes'.format(count))\n",
    "    ftp.quit()\n",
    "    logger.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "affecting-swift",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloaded genome: 470.3353 index: 0\n",
      "\n",
      "downloaded genome: 106654.48 index: 1\n",
      "\n",
      "downloaded genome: 520.659 index: 2\n",
      "\n",
      "downloaded genome: 28450.385 index: 3\n",
      "\n",
      "downloaded genome: 83554.74 index: 4\n",
      "\n",
      "downloaded genome: 813.141 index: 5\n",
      "\n",
      "downloaded genome: 545.38 index: 6\n",
      "\n",
      "downloaded genome: 777.186 index: 7\n",
      "\n",
      "downloaded genome: 1352.1760 index: 8\n",
      "\n",
      "downloaded genome: 562.22306 index: 9\n",
      "\n",
      "downloaded genome: 210.2912 index: 10\n",
      "\n",
      "downloaded genome: 573.16474 index: 11\n",
      "\n",
      "downloaded genome: 1639.2624 index: 12\n",
      "\n",
      "downloaded genome: 1041522.28 index: 13\n",
      "\n",
      "downloaded genome: 722731.3 index: 14\n",
      "\n",
      "downloaded genome: 1773.8714 index: 15\n",
      "\n",
      "downloaded genome: 2104.190 index: 16\n",
      "\n",
      "downloaded genome: 37326.9 index: 17\n",
      "\n",
      "downloaded genome: 28131.10 index: 18\n",
      "\n",
      "downloaded genome: 287.4623 index: 19\n",
      "\n",
      "downloaded genome: 1280.11681 index: 20\n",
      "\n",
      "downloaded genome: 1302.83 index: 21\n",
      "\n",
      "downloaded genome: 1338.30 index: 22\n",
      "\n",
      "downloaded genome: 1902136.3 index: 23\n",
      "\n",
      "downloaded genome: 730.54 index: 24\n",
      "\n",
      "downloaded genome: 28025.19 index: 25\n",
      "\n",
      "downloaded genome: 1496.1556 index: 26\n",
      "\n",
      "downloaded genome: 74426.49 index: 27\n",
      "\n",
      "downloaded genome: 853.172 index: 28\n",
      "\n",
      "downloaded genome: 1841863.3 index: 29\n",
      "\n",
      "downloaded genome: 1582.99 index: 30\n",
      "\n",
      "downloaded genome: 1613.133 index: 31\n",
      "\n",
      "downloaded genome: 1590.548 index: 32\n",
      "\n",
      "downloaded genome: 47715.277 index: 33\n",
      "\n",
      "downloaded genome: 487.1231 index: 34\n",
      "\n",
      "downloaded genome: 1944646.3 index: 35\n",
      "\n",
      "downloaded genome: 1849491.6 index: 36\n",
      "\n",
      "downloaded genome: 1308.247 index: 37\n",
      "\n",
      "downloaded genome: 137591.34 index: 38\n",
      "\n",
      "downloaded genome: 515619.32 index: 39\n",
      "\n",
      "downloaded genome: 470.3349 index: 40\n",
      "\n",
      "downloaded genome: 470.3350 index: 41\n",
      "\n",
      "downloaded genome: 470.3348 index: 42\n",
      "\n",
      "downloaded genome: 470.3347 index: 43\n",
      "\n",
      "downloaded genome: 470.3356 index: 44\n",
      "\n",
      "downloaded genome: 470.3357 index: 45\n",
      "\n",
      "downloaded genome: 470.3352 index: 46\n",
      "\n",
      "downloaded genome: 470.3358 index: 47\n",
      "\n",
      "downloaded genome: 470.3355 index: 48\n",
      "\n",
      "downloaded genome: 470.3351 index: 49\n",
      "\n",
      "downloaded genome: 470.3354 index: 50\n",
      "\n",
      "downloaded genome: 520.747 index: 51\n",
      "\n",
      "downloaded genome: 520.694 index: 52\n",
      "\n",
      "downloaded genome: 520.667 index: 53\n",
      "\n",
      "downloaded genome: 520.529 index: 54\n",
      "\n",
      "downloaded genome: 520.670 index: 55\n",
      "\n",
      "downloaded genome: 520.535 index: 56\n",
      "\n",
      "downloaded genome: 520.698 index: 57\n",
      "\n",
      "downloaded genome: 520.699 index: 58\n",
      "\n",
      "downloaded genome: 520.572 index: 59\n",
      "\n",
      "downloaded genome: 520.537 index: 60\n",
      "\n",
      "downloaded genome: 520.629 index: 61\n",
      "\n",
      "downloaded genome: 520.543 index: 62\n",
      "\n",
      "downloaded genome: 520.574 index: 63\n",
      "\n",
      "downloaded genome: 520.546 index: 64\n",
      "\n",
      "downloaded genome: 520.650 index: 65\n",
      "\n",
      "downloaded genome: 520.737 index: 66\n",
      "\n",
      "downloaded genome: 520.708 index: 67\n",
      "\n",
      "downloaded genome: 520.671 index: 68\n",
      "\n",
      "downloaded genome: 520.672 index: 69\n",
      "\n",
      "downloaded genome: 520.678 index: 70\n",
      "\n",
      "downloaded genome: 520.68 index: 71\n",
      "\n",
      "downloaded genome: 28450.655 index: 72\n",
      "\n",
      "downloaded genome: 813.142 index: 73\n",
      "\n",
      "downloaded genome: 168807.6 index: 74\n",
      "\n",
      "downloaded genome: 562.16466 index: 75\n",
      "\n",
      "downloaded genome: 562.22333 index: 76\n",
      "\n",
      "downloaded genome: 562.15193 index: 77\n",
      "\n",
      "downloaded genome: 562.22323 index: 78\n",
      "\n",
      "downloaded genome: 562.22307 index: 79\n",
      "\n",
      "downloaded genome: 562.16428 index: 80\n",
      "\n",
      "downloaded genome: 562.22326 index: 81\n",
      "\n",
      "downloaded genome: 573.16440 index: 82\n",
      "\n",
      "downloaded genome: 573.15319 index: 83\n",
      "\n",
      "downloaded genome: 1196172.3 index: 84\n",
      "\n",
      "downloaded genome: 1196162.3 index: 85\n",
      "\n",
      "downloaded genome: 1773.8686 index: 86\n",
      "\n",
      "downloaded genome: 1773.8719 index: 87\n",
      "\n",
      "downloaded genome: 2104.189 index: 88\n",
      "\n",
      "downloaded genome: 487.1548 index: 89\n",
      "\n",
      "downloaded genome: 287.3868 index: 90\n",
      "\n",
      "downloaded genome: 1280.12234 index: 91\n",
      "\n",
      "downloaded genome: 1280.11677 index: 92\n",
      "\n",
      "downloaded genome: 76857.43 index: 93\n",
      "\n",
      "downloaded genome: 860.11 index: 94\n",
      "\n",
      "downloaded genome: 860.17 index: 95\n",
      "\n",
      "downloaded genome: 860.16 index: 96\n",
      "\n",
      "downloaded genome: 860.18 index: 97\n",
      "\n",
      "downloaded genome: 172042.4 index: 98\n",
      "\n",
      "downloaded genome: 47715.310 index: 99\n",
      "\n",
      "downloaded 100 genomes\n"
     ]
    }
   ],
   "source": [
    "download = False\n",
    "\n",
    "output_dir_path = r'..\\..\\Data\\Bacpacs\\\\'\n",
    "\n",
    "all_genomes = df2['Genome ID']\n",
    "if download:\n",
    "    download_genomes(all_genomes, output_dir_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "german-territory",
   "metadata": {},
   "source": [
    "# Parse PATRIC genomes files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "registered-western",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacpacs_dir_path = r'..\\..\\Data\\Bacpacs\\\\'\n",
    "bacpacs_patric_files_path = bacpacs_dir_path + 'patric_files\\\\' \n",
    "bacpacs_genomes_path = bacpacs_dir_path + 'bacpacs_genomes.fasta'\n",
    "bacpacs_metadata_path = bacpacs_dir_path + 'bacpacs_test.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "permanent-petite",
   "metadata": {},
   "outputs": [],
   "source": [
    "UNKNOWN_PGFAM = 'X'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "available-audit",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_patric_file(file_path):\n",
    "    \n",
    "    genome_df = pd.read_csv(file_path, sep='\\t')\n",
    "    pgfams = genome_df['pgfam_id'].fillna(UNKNOWN_PGFAM)\n",
    "    \n",
    "    return pgfams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "creative-reference",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_genomes_files(patric_files_folder, output_path)\n",
    "    with os.scandir(patric_files_folder) as entries:\n",
    "        with open(output_path, 'w') as out_f:\n",
    "            for entry in entries:\n",
    "                if entry.is_file():\n",
    "                    genome_id = entry.name.split(FILE_SUFFIX)[0]\n",
    "                    genome_pgfams = parse_patric_file(entry)\n",
    "\n",
    "                    out_f.write(f'>{genome_id}\\n')\n",
    "                    out_f.write('\\n'.join(genome_pgfams))\n",
    "                    out_f.write('\\n')\n",
    "                    \n",
    "#write_genomes_files(bacpacs_patric_files_path, bacpacs_genomes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charitable-albuquerque",
   "metadata": {},
   "source": [
    "# Load Bacpacs Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "executed-organ",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacpacs_dataset = GenomesData(bacpacs_genomes_path, bacpacs_metadata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "certain-chassis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bacpacs_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "psychological-battlefield",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>BacPaCS Label</th>\n",
       "      <th>Label</th>\n",
       "      <th>Ref.</th>\n",
       "      <th>OPP</th>\n",
       "      <th>Balanced Test</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>470.3353</th>\n",
       "      <td>1</td>\n",
       "      <td>Acinetobacter baumannii strain HWBA8</td>\n",
       "      <td>NHP</td>\n",
       "      <td>HP</td>\n",
       "      <td>Clinical isolate \\cite{yoon2017bla}</td>\n",
       "      <td>\\cite{peleg2008acinetobacter}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106654.48</th>\n",
       "      <td>2</td>\n",
       "      <td>Acinetobacter nosocomialis strain SSA3</td>\n",
       "      <td>NHP</td>\n",
       "      <td>HP</td>\n",
       "      <td>Clinical isolate \\cite{yoon2017bla}</td>\n",
       "      <td>\\cite{peleg2008acinetobacter}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520.659</th>\n",
       "      <td>3</td>\n",
       "      <td>Bordetella pertussis strain B227</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{bowden2016genome}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28450.385</th>\n",
       "      <td>4</td>\n",
       "      <td>Burkholderia pseudomallei strain MSHR5864</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{sarovich2018raising}</td>\n",
       "      <td>\\cite{mangalea2017nitrate}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83554.74</th>\n",
       "      <td>5</td>\n",
       "      <td>Chlamydia psittaci strain GIMC 2005:CpsCP1</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{feodorova2020data}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487.1548</th>\n",
       "      <td>90</td>\n",
       "      <td>Neisseria meningitidis strain M26417</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{folaranmi2017increased}</td>\n",
       "      <td>\\cite{bernardini2004proteome}</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287.3868</th>\n",
       "      <td>91</td>\n",
       "      <td>Pseudomonas aeruginosa strain RIVM-EMC2982</td>\n",
       "      <td>NHP</td>\n",
       "      <td>HP</td>\n",
       "      <td>Clinical isolate \\cite{botelho2018two}</td>\n",
       "      <td>\\cite{diggle2020microbe}</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280.12234</th>\n",
       "      <td>92</td>\n",
       "      <td>Staphylococcus aureus strain JE2</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{1280.12234,o2017into}</td>\n",
       "      <td>\\cite{o2017into}</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280.11677</th>\n",
       "      <td>93</td>\n",
       "      <td>Staphylococcus aureus strain USA300-SUR11</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{1280.11677,o2017into}</td>\n",
       "      <td>\\cite{o2017into}</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47715.310</th>\n",
       "      <td>100</td>\n",
       "      <td>Lactobacillus rhamnosus strain LR5</td>\n",
       "      <td>NHP</td>\n",
       "      <td>NHP</td>\n",
       "      <td>\\cite{gao2019novel}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Unnamed: 0                                 Genome Name  \\\n",
       "Genome ID                                                           \n",
       "470.3353            1        Acinetobacter baumannii strain HWBA8   \n",
       "106654.48           2      Acinetobacter nosocomialis strain SSA3   \n",
       "520.659             3            Bordetella pertussis strain B227   \n",
       "28450.385           4   Burkholderia pseudomallei strain MSHR5864   \n",
       "83554.74            5  Chlamydia psittaci strain GIMC 2005:CpsCP1   \n",
       "...               ...                                         ...   \n",
       "487.1548           90        Neisseria meningitidis strain M26417   \n",
       "287.3868           91  Pseudomonas aeruginosa strain RIVM-EMC2982   \n",
       "1280.12234         92            Staphylococcus aureus strain JE2   \n",
       "1280.11677         93   Staphylococcus aureus strain USA300-SUR11   \n",
       "47715.310         100          Lactobacillus rhamnosus strain LR5   \n",
       "\n",
       "           BacPaCS Label Label                                    Ref.  \\\n",
       "Genome ID                                                                \n",
       "470.3353             NHP    HP     Clinical isolate \\cite{yoon2017bla}   \n",
       "106654.48            NHP    HP     Clinical isolate \\cite{yoon2017bla}   \n",
       "520.659               HP    HP                 \\cite{bowden2016genome}   \n",
       "28450.385             HP    HP              \\cite{sarovich2018raising}   \n",
       "83554.74              HP    HP                \\cite{feodorova2020data}   \n",
       "...                  ...   ...                                     ...   \n",
       "487.1548              HP    HP           \\cite{folaranmi2017increased}   \n",
       "287.3868             NHP    HP  Clinical isolate \\cite{botelho2018two}   \n",
       "1280.12234            HP    HP             \\cite{1280.12234,o2017into}   \n",
       "1280.11677            HP    HP             \\cite{1280.11677,o2017into}   \n",
       "47715.310            NHP   NHP                     \\cite{gao2019novel}   \n",
       "\n",
       "                                      OPP Balanced Test  \n",
       "Genome ID                                                \n",
       "470.3353    \\cite{peleg2008acinetobacter}           Yes  \n",
       "106654.48   \\cite{peleg2008acinetobacter}           Yes  \n",
       "520.659                               NaN           Yes  \n",
       "28450.385      \\cite{mangalea2017nitrate}           Yes  \n",
       "83554.74                              NaN           Yes  \n",
       "...                                   ...           ...  \n",
       "487.1548    \\cite{bernardini2004proteome}            No  \n",
       "287.3868         \\cite{diggle2020microbe}            No  \n",
       "1280.12234               \\cite{o2017into}            No  \n",
       "1280.11677               \\cite{o2017into}            No  \n",
       "47715.310                             NaN            No  \n",
       "\n",
       "[94 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacpacs_dataset.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dressed-tamil",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = bacpacs_dataset.y"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stable-schedule",
   "metadata": {},
   "source": [
    "## Balanced dataset (one genome per species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "indie-killer",
   "metadata": {},
   "outputs": [],
   "source": [
    "balanced_dataset_genomes = bacpacs_dataset.metadata[bacpacs_dataset.metadata['Balanced Test'] == 'Yes'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "healthy-bandwidth",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacpacs_data_balanced = bacpacs_dataset.data[balanced_dataset_genomes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "norman-arkansas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bacpacs_data_balanced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "external-sleeve",
   "metadata": {},
   "outputs": [],
   "source": [
    "balanced_labels = labels[balanced_dataset_genomes]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vocational-interface",
   "metadata": {},
   "source": [
    "# WSPC model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "junior-jewelry",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = '..\\..\\WSPC Model\\WSPC_model.pkl'\n",
    "\n",
    "WSPC = util.load_model(model_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "freelance-delivery",
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = WSPC.predict(bacpacs_dataset.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "steady-rings",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "         NHP       0.93      0.88      0.90        16\n",
      "          HP       0.97      0.99      0.98        78\n",
      "\n",
      "    accuracy                           0.97        94\n",
      "   macro avg       0.95      0.93      0.94        94\n",
      "weighted avg       0.97      0.97      0.97        94\n",
      "\n"
     ]
    }
   ],
   "source": [
    "target_names = ['NHP', 'HP']\n",
    "print(classification_report(labels, preds, target_names=target_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bored-safety",
   "metadata": {},
   "source": [
    "NHP recall = specificity   \n",
    "HP recall = sensitivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "opposed-strategy",
   "metadata": {},
   "outputs": [],
   "source": [
    "balanced_preds = WSPC.predict(bacpacs_data_balanced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "monetary-assembly",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "         NHP       1.00      0.87      0.93        15\n",
      "          HP       0.93      1.00      0.96        25\n",
      "\n",
      "    accuracy                           0.95        40\n",
      "   macro avg       0.96      0.93      0.95        40\n",
      "weighted avg       0.95      0.95      0.95        40\n",
      "\n"
     ]
    }
   ],
   "source": [
    "target_names = ['NHP', 'HP']\n",
    "print(classification_report(balanced_labels, balanced_preds, target_names=target_names))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
