{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "suitable-likelihood",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys  \n",
    "sys.path.insert(0, '../../Code')\n",
    "\n",
    "import os\n",
    "import astropy\n",
    "from astropy.table import Table\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from ftplib import FTP\n",
    "\n",
    "import train_and_predict\n",
    "import util\n",
    "from data import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "composed-restoration",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "convertible-faith",
   "metadata": {},
   "outputs": [],
   "source": [
    "download = False\n",
    "write = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "amber-county",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacpacs_dir_path = r'..\\..\\Data\\Bacpacs\\\\'\n",
    "bacpacs_patric_files_path = bacpacs_dir_path + 'patric_files\\\\' \n",
    "bacpacs_genomes_path = bacpacs_dir_path + 'bacpacs_genomes.fasta'\n",
    "bacpacs_metadata_path = bacpacs_dir_path + 'bacpacs_test.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "trying-shoulder",
   "metadata": {},
   "source": [
    "read latex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "cathedral-frost",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<i>Table length=100</i>\n",
       "<table id=\"table2981039572040\" class=\"table-striped table-bordered table-condensed\">\n",
       "<thead><tr><th>col0</th><th>Genome ID</th><th>Genome Name</th><th>BacPaCS Label</th><th>Validated Label</th><th>Ref.</th><th>OPP</th><th>Balanced Test</th></tr></thead>\n",
       "<thead><tr><th>int32</th><th>str10</th><th>str64</th><th>str3</th><th>str3</th><th>str61</th><th>str31</th><th>str3</th></tr></thead>\n",
       "<tr><td>1</td><td>470.3353</td><td>Acinetobacter baumannii strain HWBA8</td><td>NHP</td><td>HP</td><td>Clinical isolate \\cite{yoon2017bla}</td><td>\\cite{peleg2008acinetobacter}</td><td>Yes</td></tr>\n",
       "<tr><td>2</td><td>106654.48</td><td>Acinetobacter nosocomialis strain SSA3</td><td>NHP</td><td>HP</td><td>Clinical isolate \\cite{yoon2017bla}</td><td>\\cite{peleg2008acinetobacter}</td><td>Yes</td></tr>\n",
       "<tr><td>3</td><td>520.659</td><td>Bordetella pertussis strain B227</td><td>HP</td><td>HP</td><td>\\cite{bowden2016genome}</td><td>--</td><td>Yes</td></tr>\n",
       "<tr><td>4</td><td>28450.385</td><td>Burkholderia pseudomallei strain MSHR5864</td><td>HP</td><td>HP</td><td>\\cite{sarovich2018raising}</td><td>\\cite{mangalea2017nitrate}</td><td>Yes</td></tr>\n",
       "<tr><td>5</td><td>83554.74</td><td>Chlamydia psittaci strain GIMC 2005:CpsCP1</td><td>HP</td><td>HP</td><td>\\cite{feodorova2020data}</td><td>--</td><td>Yes</td></tr>\n",
       "<tr><td>6</td><td>813.141</td><td>Chlamydia trachomatis strain SQ12</td><td>HP</td><td>HP</td><td>\\cite{suchland2017demonstration}</td><td>--</td><td>Yes</td></tr>\n",
       "<tr><td>7</td><td>545.38</td><td>Citrobacter koseri strain FDAARGOS\\_287</td><td>HP</td><td>HP</td><td>\\cite{545.38,yuan2019comparative}</td><td>\\cite{yuan2019comparative}</td><td>Yes</td></tr>\n",
       "<tr><td>8</td><td>777.186</td><td>Coxiella burnetii strain Heizberg</td><td>HP</td><td>HP</td><td>\\cite{kuley2017genome}</td><td>--</td><td>Yes</td></tr>\n",
       "<tr><td>9</td><td>1352.1760</td><td>Enterococcus faecium strain Ef\\_aus00233</td><td>NHP</td><td>HP</td><td>Hospital outbreak isolate \\cite{buultjens2017evolutionary}</td><td>\\cite{zhou2020enterococcus}</td><td>Yes</td></tr>\n",
       "<tr><td>10</td><td>562.22306</td><td>Escherichia coli strain FDAARGOS\\_433 strain Not applicable</td><td>HP</td><td>HP</td><td>\\cite{562.22306}</td><td>--</td><td>Yes</td></tr>\n",
       "<tr><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "<tr><td>91</td><td>287.3868</td><td>Pseudomonas aeruginosa strain RIVM-EMC2982</td><td>NHP</td><td>HP</td><td>Clinical isolate \\cite{botelho2018two}</td><td>\\cite{diggle2020microbe}</td><td>No</td></tr>\n",
       "<tr><td>92</td><td>1280.12234</td><td>Staphylococcus aureus strain JE2</td><td>HP</td><td>HP</td><td>\\cite{1280.12234,o2017into}</td><td>\\cite{o2017into}</td><td>No</td></tr>\n",
       "<tr><td>93</td><td>1280.11677</td><td>Staphylococcus aureus strain USA300-SUR11</td><td>HP</td><td>HP</td><td>\\cite{1280.11677,o2017into}</td><td>\\cite{o2017into}</td><td>No</td></tr>\n",
       "<tr><td>94</td><td>76857.43</td><td>Fusobacterium nucleatum subsp. polymorphum strain KCOM 1275</td><td>NHP</td><td>INC</td><td>\\cite{han2015fusobacterium}</td><td>--</td><td>No</td></tr>\n",
       "<tr><td>95</td><td>172042.4</td><td>Rothia aeria strain JCM 11412</td><td>NHP</td><td>INC</td><td>--</td><td>\\cite{nambu2016complete}</td><td>No</td></tr>\n",
       "<tr><td>96</td><td>860.11</td><td>Fusobacterium periodonticum strain KCOM 1261</td><td>NHP</td><td>INC</td><td>\\cite{sphaerophorus2006genus}</td><td>--</td><td>No</td></tr>\n",
       "<tr><td>97</td><td>860.17</td><td>Fusobacterium periodonticum strain KCOM 1262</td><td>NHP</td><td>INC</td><td>\\cite{sphaerophorus2006genus}</td><td>--</td><td>No</td></tr>\n",
       "<tr><td>98</td><td>860.16</td><td>Fusobacterium periodonticum strain KCOM 2555</td><td>NHP</td><td>INC</td><td>\\cite{sphaerophorus2006genus}</td><td>--</td><td>No</td></tr>\n",
       "<tr><td>99</td><td>860.18</td><td>Fusobacterium periodonticum strain KCOM 2653</td><td>NHP</td><td>INC</td><td>\\cite{sphaerophorus2006genus}</td><td>--</td><td>No</td></tr>\n",
       "<tr><td>100</td><td>47715.310</td><td>Lactobacillus rhamnosus strain LR5</td><td>NHP</td><td>NHP</td><td>\\cite{gao2019novel}</td><td>--</td><td>No</td></tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<Table length=100>\n",
       " col0 Genome ID  ...              OPP              Balanced Test\n",
       "int32   str10    ...             str31                  str3    \n",
       "----- ---------- ... ----------------------------- -------------\n",
       "    1   470.3353 ... \\cite{peleg2008acinetobacter}           Yes\n",
       "    2  106654.48 ... \\cite{peleg2008acinetobacter}           Yes\n",
       "    3    520.659 ...                            --           Yes\n",
       "    4  28450.385 ...    \\cite{mangalea2017nitrate}           Yes\n",
       "    5   83554.74 ...                            --           Yes\n",
       "    6    813.141 ...                            --           Yes\n",
       "    7     545.38 ...    \\cite{yuan2019comparative}           Yes\n",
       "    8    777.186 ...                            --           Yes\n",
       "    9  1352.1760 ...   \\cite{zhou2020enterococcus}           Yes\n",
       "   10  562.22306 ...                            --           Yes\n",
       "  ...        ... ...                           ...           ...\n",
       "   91   287.3868 ...      \\cite{diggle2020microbe}            No\n",
       "   92 1280.12234 ...              \\cite{o2017into}            No\n",
       "   93 1280.11677 ...              \\cite{o2017into}            No\n",
       "   94   76857.43 ...                            --            No\n",
       "   95   172042.4 ...      \\cite{nambu2016complete}            No\n",
       "   96     860.11 ...                            --            No\n",
       "   97     860.17 ...                            --            No\n",
       "   98     860.16 ...                            --            No\n",
       "   99     860.18 ...                            --            No\n",
       "  100  47715.310 ...                            --            No"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = r'..\\..\\Paper\\Tables\\\\'\n",
    "\n",
    "converters = {'Genome ID': [astropy.io.ascii.convert_numpy(np.str)]}\n",
    "tab = Table.read(path+'test_set_bacpacs.tex', converters=converters)\n",
    "tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mounted-miami",
   "metadata": {},
   "source": [
    "convert to pandas DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "radio-oxygen",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>BacPaCS Label</th>\n",
       "      <th>Validated Label</th>\n",
       "      <th>Ref.</th>\n",
       "      <th>OPP</th>\n",
       "      <th>Balanced Test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>470.3353</td>\n",
       "      <td>Acinetobacter baumannii strain HWBA8</td>\n",
       "      <td>NHP</td>\n",
       "      <td>HP</td>\n",
       "      <td>Clinical isolate \\cite{yoon2017bla}</td>\n",
       "      <td>\\cite{peleg2008acinetobacter}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>106654.48</td>\n",
       "      <td>Acinetobacter nosocomialis strain SSA3</td>\n",
       "      <td>NHP</td>\n",
       "      <td>HP</td>\n",
       "      <td>Clinical isolate \\cite{yoon2017bla}</td>\n",
       "      <td>\\cite{peleg2008acinetobacter}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>520.659</td>\n",
       "      <td>Bordetella pertussis strain B227</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{bowden2016genome}</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28450.385</td>\n",
       "      <td>Burkholderia pseudomallei strain MSHR5864</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{sarovich2018raising}</td>\n",
       "      <td>\\cite{mangalea2017nitrate}</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>83554.74</td>\n",
       "      <td>Chlamydia psittaci strain GIMC 2005:CpsCP1</td>\n",
       "      <td>HP</td>\n",
       "      <td>HP</td>\n",
       "      <td>\\cite{feodorova2020data}</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>860.17</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 1262</td>\n",
       "      <td>NHP</td>\n",
       "      <td>INC</td>\n",
       "      <td>\\cite{sphaerophorus2006genus}</td>\n",
       "      <td></td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>860.16</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 2555</td>\n",
       "      <td>NHP</td>\n",
       "      <td>INC</td>\n",
       "      <td>\\cite{sphaerophorus2006genus}</td>\n",
       "      <td></td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>860.18</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 2653</td>\n",
       "      <td>NHP</td>\n",
       "      <td>INC</td>\n",
       "      <td>\\cite{sphaerophorus2006genus}</td>\n",
       "      <td></td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>172042.4</td>\n",
       "      <td>Rothia aeria strain JCM 11412</td>\n",
       "      <td>NHP</td>\n",
       "      <td>INC</td>\n",
       "      <td></td>\n",
       "      <td>\\cite{nambu2016complete}</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>47715.310</td>\n",
       "      <td>Lactobacillus rhamnosus strain LR5</td>\n",
       "      <td>NHP</td>\n",
       "      <td>NHP</td>\n",
       "      <td>\\cite{gao2019novel}</td>\n",
       "      <td></td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Genome ID                                   Genome Name BacPaCS Label  \\\n",
       "1     470.3353          Acinetobacter baumannii strain HWBA8           NHP   \n",
       "2    106654.48        Acinetobacter nosocomialis strain SSA3           NHP   \n",
       "3      520.659              Bordetella pertussis strain B227            HP   \n",
       "4    28450.385     Burkholderia pseudomallei strain MSHR5864            HP   \n",
       "5     83554.74    Chlamydia psittaci strain GIMC 2005:CpsCP1            HP   \n",
       "..         ...                                           ...           ...   \n",
       "96      860.17  Fusobacterium periodonticum strain KCOM 1262           NHP   \n",
       "97      860.16  Fusobacterium periodonticum strain KCOM 2555           NHP   \n",
       "98      860.18  Fusobacterium periodonticum strain KCOM 2653           NHP   \n",
       "99    172042.4                 Rothia aeria strain JCM 11412           NHP   \n",
       "100  47715.310            Lactobacillus rhamnosus strain LR5           NHP   \n",
       "\n",
       "    Validated Label                                 Ref.  \\\n",
       "1                HP  Clinical isolate \\cite{yoon2017bla}   \n",
       "2                HP  Clinical isolate \\cite{yoon2017bla}   \n",
       "3                HP              \\cite{bowden2016genome}   \n",
       "4                HP           \\cite{sarovich2018raising}   \n",
       "5                HP             \\cite{feodorova2020data}   \n",
       "..              ...                                  ...   \n",
       "96              INC        \\cite{sphaerophorus2006genus}   \n",
       "97              INC        \\cite{sphaerophorus2006genus}   \n",
       "98              INC        \\cite{sphaerophorus2006genus}   \n",
       "99              INC                                        \n",
       "100             NHP                  \\cite{gao2019novel}   \n",
       "\n",
       "                               OPP Balanced Test  \n",
       "1    \\cite{peleg2008acinetobacter}           Yes  \n",
       "2    \\cite{peleg2008acinetobacter}           Yes  \n",
       "3                                            Yes  \n",
       "4       \\cite{mangalea2017nitrate}           Yes  \n",
       "5                                            Yes  \n",
       "..                             ...           ...  \n",
       "96                                            No  \n",
       "97                                            No  \n",
       "98                                            No  \n",
       "99        \\cite{nambu2016complete}            No  \n",
       "100                                           No  \n",
       "\n",
       "[100 rows x 7 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacpacs_tex_df = tab.to_pandas()\n",
    "bacpacs_tex_df = bacpacs_tex_df.drop('col0', axis=1)\n",
    "bacpacs_tex_df = bacpacs_tex_df.sort_values(['Balanced Test', 'Validated Label', 'Genome Name'], ascending=[False, True, True])\n",
    "bacpacs_tex_df.index = range(1, len(bacpacs_tex_df)+1)\n",
    "bacpacs_tex_df = bacpacs_tex_df.fillna('')\n",
    "\n",
    "bacpacs_tex_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "corresponding-soundtrack",
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME_ID = 'Genome ID'\n",
    "\n",
    "bacpacs_sup = pd.read_csv(bacpacs_dir_path + 'Bacpacs-Supplementary.csv', dtype='str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "accurate-portuguese",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>pathogenicity</th>\n",
       "      <th>bacpacs</th>\n",
       "      <th>pathogenfinder</th>\n",
       "      <th>paprbag_fold1</th>\n",
       "      <th>paprbag_fold2</th>\n",
       "      <th>paprbag_fold3</th>\n",
       "      <th>paprbag_fold4</th>\n",
       "      <th>paprbag_fold5</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Organism Name</th>\n",
       "      <th>Isolation Source</th>\n",
       "      <th>Habitat</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Host Health</th>\n",
       "      <th>Other Clinical</th>\n",
       "      <th>Comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1041522.28</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Mycobacterium colombiense CECT 3035 strain CEC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M. colombiense infections were initially descr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1196162.3</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Listeria monocytogenes serotype 4b str. 10-080...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Whole genome sequencing and comparison of List...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1196172.3</td>\n",
       "      <td>hp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Listeria monocytogenes serotype 4b str. 02-128...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Whole genome sequencing and comparison of List...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1280.11681</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Staphylococcus aureus strain USA300-SUR15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The study characterizes a Staphylococcus aureu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1280.12234</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Staphylococcus aureus strain JE2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>skin and soft tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soft Tissue Infections</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sequence the genome of Staphylococcus aureus JE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>853.172</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>Faecalibacterium prausnitzii strain Indica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Faecalibacterium prausnitzii is the most abund...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>860.11</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 1261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tongue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genome sequencing of Fusobacterium periodontic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>860.16</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 2555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dental plaque</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genome sequencing of Fusobacterium periodontic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>860.17</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 1262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tongue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genome sequencing of Fusobacterium periodontic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>860.18</td>\n",
       "      <td>nhp</td>\n",
       "      <td>nhp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>hp</td>\n",
       "      <td>Fusobacterium periodonticum strain KCOM 2653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tongue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genome sequencing of Fusobacterium periodontic...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Genome ID pathogenicity bacpacs pathogenfinder paprbag_fold1  \\\n",
       "0   1041522.28            hp      hp             hp            hp   \n",
       "1    1196162.3            hp      hp             hp            hp   \n",
       "2    1196172.3            hp     nhp             hp            hp   \n",
       "3   1280.11681            hp      hp             hp            hp   \n",
       "4   1280.12234            hp      hp             hp            hp   \n",
       "..         ...           ...     ...            ...           ...   \n",
       "95     853.172           nhp     nhp            nhp           nhp   \n",
       "96      860.11           nhp     nhp             hp            hp   \n",
       "97      860.16           nhp     nhp             hp            hp   \n",
       "98      860.17           nhp     nhp             hp            hp   \n",
       "99      860.18           nhp     nhp             hp            hp   \n",
       "\n",
       "   paprbag_fold2 paprbag_fold3 paprbag_fold4 paprbag_fold5  \\\n",
       "0             hp            hp            hp            hp   \n",
       "1             hp            hp            hp            hp   \n",
       "2             hp            hp            hp            hp   \n",
       "3             hp            hp            hp            hp   \n",
       "4             hp            hp            hp            hp   \n",
       "..           ...           ...           ...           ...   \n",
       "95           nhp           nhp           nhp           nhp   \n",
       "96            hp            hp            hp            hp   \n",
       "97            hp            hp            hp            hp   \n",
       "98            hp            hp            hp            hp   \n",
       "99            hp            hp            hp            hp   \n",
       "\n",
       "                                          Genome Name Organism Name  \\\n",
       "0   Mycobacterium colombiense CECT 3035 strain CEC...           NaN   \n",
       "1   Listeria monocytogenes serotype 4b str. 10-080...           NaN   \n",
       "2   Listeria monocytogenes serotype 4b str. 02-128...           NaN   \n",
       "3           Staphylococcus aureus strain USA300-SUR15           NaN   \n",
       "4                    Staphylococcus aureus strain JE2           NaN   \n",
       "..                                                ...           ...   \n",
       "95         Faecalibacterium prausnitzii strain Indica           NaN   \n",
       "96       Fusobacterium periodonticum strain KCOM 1261           NaN   \n",
       "97       Fusobacterium periodonticum strain KCOM 2555           NaN   \n",
       "98       Fusobacterium periodonticum strain KCOM 1262           NaN   \n",
       "99       Fusobacterium periodonticum strain KCOM 2653           NaN   \n",
       "\n",
       "        Isolation Source Habitat Disease             Host Health  \\\n",
       "0                  blood     NaN     NaN                     HIV   \n",
       "1                  Stool     NaN     NaN             Listeriosis   \n",
       "2                  Stool     NaN     NaN             Listeriosis   \n",
       "3                   Nose     NaN     NaN                     NaN   \n",
       "4   skin and soft tissue     NaN     NaN  Soft Tissue Infections   \n",
       "..                   ...     ...     ...                     ...   \n",
       "95                   NaN     NaN     NaN                     NaN   \n",
       "96                Tongue     NaN     NaN                  normal   \n",
       "97         Dental plaque     NaN     NaN                  normal   \n",
       "98                Tongue     NaN     NaN                  normal   \n",
       "99                Tongue     NaN     NaN                  normal   \n",
       "\n",
       "   Other Clinical                                           Comments  \n",
       "0             NaN  M. colombiense infections were initially descr...  \n",
       "1             NaN  Whole genome sequencing and comparison of List...  \n",
       "2             NaN  Whole genome sequencing and comparison of List...  \n",
       "3             NaN  The study characterizes a Staphylococcus aureu...  \n",
       "4             NaN   Sequence the genome of Staphylococcus aureus JE2  \n",
       "..            ...                                                ...  \n",
       "95            NaN  Faecalibacterium prausnitzii is the most abund...  \n",
       "96            NaN  Genome sequencing of Fusobacterium periodontic...  \n",
       "97            NaN  Genome sequencing of Fusobacterium periodontic...  \n",
       "98            NaN  Genome sequencing of Fusobacterium periodontic...  \n",
       "99            NaN  Genome sequencing of Fusobacterium periodontic...  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacpacs_sup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "incredible-cylinder",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(bacpacs_sup[GENOME_ID]) - set(bacpacs_tex_df[GENOME_ID])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ignored-kinase",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacpacs_metadata = bacpacs_tex_df.rename(columns={'Validated Label': 'Label'})\n",
    "bacpacs_metadata = bacpacs_metadata[bacpacs_metadata['Label'] != 'INC']\n",
    "bacpacs_metadata.to_csv(bacpacs_metadata_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "light-roots",
   "metadata": {},
   "source": [
    "# Download Bacpacs Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "historical-highlight",
   "metadata": {},
   "outputs": [],
   "source": [
    "FILE_SUFFIX = '.PATRIC.features.tab'\n",
    "logger = open('logger', 'w')\n",
    "\n",
    "def read_genome_ids(input_file):\n",
    "    with open(input_file) as f:\n",
    "        ids = f.read().strip().split('\\n')\n",
    "    return ids\n",
    "\n",
    "def grab_file(ftp, filename, local_path):\n",
    "\n",
    "    local_file = open(local_path, 'wb')\n",
    "\n",
    "    try:\n",
    "        ftp.retrbinary('RETR ' + filename, local_file.write, 1024)\n",
    "    except Exception as e:\n",
    "        logger.write(e)\n",
    "\n",
    "    local_file.close()\n",
    "\n",
    "\n",
    "def download_genomes(genomes, output_dir_path):\n",
    "\n",
    "\n",
    "    # connect to host, default port\n",
    "    ftp = FTP(r'ftp.patricbrc.org')\n",
    "\n",
    "    ftp.login()  # user anonymous, passwd anonymous@\n",
    "\n",
    "    count = 0\n",
    "    for i, genome_id in enumerate(genomes):\n",
    "\n",
    "        try:\n",
    "            ftp.cwd(\"/genomes/\" + genome_id)\n",
    "            local_file = output_dir_path + genome_id + FILE_SUFFIX\n",
    "            remote_file = genome_id + FILE_SUFFIX\n",
    "            grab_file(ftp, remote_file, local_file)\n",
    "\n",
    "            print('downloaded genome: ' + genome_id + ' index: ' + str(i) + '\\n')\n",
    "            count += 1\n",
    "\n",
    "        except Exception as e:\n",
    "            # print('ERROR genome: ' + genome_id + ' index: ' + str(i) + '\\n')\n",
    "            logger.write(f'genome: {genome_id} index: {i} error: {e}\\n')\n",
    "\n",
    "    print('downloaded {} genomes'.format(count))\n",
    "    ftp.quit()\n",
    "    logger.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "affecting-swift",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloaded genome: 1041522.28 index: 0\n",
      "\n",
      "downloaded genome: 1196162.3 index: 1\n",
      "\n",
      "downloaded genome: 1196172.3 index: 2\n",
      "\n",
      "downloaded genome: 1280.11681 index: 3\n",
      "\n",
      "downloaded genome: 1280.12234 index: 4\n",
      "\n",
      "downloaded genome: 1302.83 index: 5\n",
      "\n",
      "downloaded genome: 1338.30 index: 6\n",
      "\n",
      "downloaded genome: 1639.2624 index: 7\n",
      "\n",
      "downloaded genome: 1773.8686 index: 8\n",
      "\n",
      "downloaded genome: 1773.8714 index: 9\n",
      "\n",
      "downloaded genome: 1773.8719 index: 10\n",
      "\n",
      "downloaded genome: 210.2912 index: 11\n",
      "\n",
      "downloaded genome: 2104.189 index: 12\n",
      "\n",
      "downloaded genome: 2104.190 index: 13\n",
      "\n",
      "downloaded genome: 28131.10 index: 14\n",
      "\n",
      "downloaded genome: 28450.655 index: 15\n",
      "\n",
      "downloaded genome: 287.4623 index: 16\n",
      "\n",
      "downloaded genome: 37326.9 index: 17\n",
      "\n",
      "downloaded genome: 487.1548 index: 18\n",
      "\n",
      "downloaded genome: 520.529 index: 19\n",
      "\n",
      "downloaded genome: 520.535 index: 20\n",
      "\n",
      "downloaded genome: 520.537 index: 21\n",
      "\n",
      "downloaded genome: 520.543 index: 22\n",
      "\n",
      "downloaded genome: 520.546 index: 23\n",
      "\n",
      "downloaded genome: 520.572 index: 24\n",
      "\n",
      "downloaded genome: 520.629 index: 25\n",
      "\n",
      "downloaded genome: 520.650 index: 26\n",
      "\n",
      "downloaded genome: 520.659 index: 27\n",
      "\n",
      "downloaded genome: 520.667 index: 28\n",
      "\n",
      "downloaded genome: 520.670 index: 29\n",
      "\n",
      "downloaded genome: 520.671 index: 30\n",
      "\n",
      "downloaded genome: 520.672 index: 31\n",
      "\n",
      "downloaded genome: 520.678 index: 32\n",
      "\n",
      "downloaded genome: 520.680 index: 33\n",
      "\n",
      "downloaded genome: 520.694 index: 34\n",
      "\n",
      "downloaded genome: 520.698 index: 35\n",
      "\n",
      "downloaded genome: 520.699 index: 36\n",
      "\n",
      "downloaded genome: 520.708 index: 37\n",
      "\n",
      "downloaded genome: 520.737 index: 38\n",
      "\n",
      "downloaded genome: 520.747 index: 39\n",
      "\n",
      "downloaded genome: 545.38 index: 40\n",
      "\n",
      "downloaded genome: 562.15193 index: 41\n",
      "\n",
      "downloaded genome: 562.16466 index: 42\n",
      "\n",
      "downloaded genome: 562.22306 index: 43\n",
      "\n",
      "downloaded genome: 562.22307 index: 44\n",
      "\n",
      "downloaded genome: 562.22323 index: 45\n",
      "\n",
      "downloaded genome: 562.22326 index: 46\n",
      "\n",
      "downloaded genome: 562.22333 index: 47\n",
      "\n",
      "downloaded genome: 573.15319 index: 48\n",
      "\n",
      "downloaded genome: 573.16440 index: 49\n",
      "\n",
      "downloaded genome: 573.16474 index: 50\n",
      "\n",
      "downloaded genome: 722731.3 index: 51\n",
      "\n",
      "downloaded genome: 730.54 index: 52\n",
      "\n",
      "downloaded genome: 777.186 index: 53\n",
      "\n",
      "downloaded genome: 813.141 index: 54\n",
      "\n",
      "downloaded genome: 813.142 index: 55\n",
      "\n",
      "downloaded genome: 1280.11677 index: 56\n",
      "\n",
      "downloaded genome: 28450.385 index: 57\n",
      "\n",
      "downloaded genome: 520.574 index: 58\n",
      "\n",
      "downloaded genome: 83554.74 index: 59\n",
      "\n",
      "downloaded genome: 106654.48 index: 60\n",
      "\n",
      "downloaded genome: 1308.247 index: 61\n",
      "\n",
      "downloaded genome: 1352.1760 index: 62\n",
      "\n",
      "downloaded genome: 137591.34 index: 63\n",
      "\n",
      "downloaded genome: 1496.1556 index: 64\n",
      "\n",
      "downloaded genome: 1582.99 index: 65\n",
      "\n",
      "downloaded genome: 1590.548 index: 66\n",
      "\n",
      "downloaded genome: 1613.133 index: 67\n",
      "\n",
      "downloaded genome: 168807.6 index: 68\n",
      "\n",
      "downloaded genome: 172042.4 index: 69\n",
      "\n",
      "downloaded genome: 1841863.3 index: 70\n",
      "\n",
      "downloaded genome: 1849491.6 index: 71\n",
      "\n",
      "downloaded genome: 1902136.3 index: 72\n",
      "\n",
      "downloaded genome: 1944646.3 index: 73\n",
      "\n",
      "downloaded genome: 28025.19 index: 74\n",
      "\n",
      "downloaded genome: 287.3868 index: 75\n",
      "\n",
      "downloaded genome: 470.3347 index: 76\n",
      "\n",
      "downloaded genome: 470.3348 index: 77\n",
      "\n",
      "downloaded genome: 470.3349 index: 78\n",
      "\n",
      "downloaded genome: 470.3350 index: 79\n",
      "\n",
      "downloaded genome: 470.3351 index: 80\n",
      "\n",
      "downloaded genome: 470.3352 index: 81\n",
      "\n",
      "downloaded genome: 470.3353 index: 82\n",
      "\n",
      "downloaded genome: 470.3354 index: 83\n",
      "\n",
      "downloaded genome: 470.3355 index: 84\n",
      "\n",
      "downloaded genome: 470.3356 index: 85\n",
      "\n",
      "downloaded genome: 470.3357 index: 86\n",
      "\n",
      "downloaded genome: 470.3358 index: 87\n",
      "\n",
      "downloaded genome: 47715.277 index: 88\n",
      "\n",
      "downloaded genome: 47715.310 index: 89\n",
      "\n",
      "downloaded genome: 487.1231 index: 90\n",
      "\n",
      "downloaded genome: 515619.32 index: 91\n",
      "\n",
      "downloaded genome: 562.16428 index: 92\n",
      "\n",
      "downloaded genome: 74426.49 index: 93\n",
      "\n",
      "downloaded genome: 76857.43 index: 94\n",
      "\n",
      "downloaded genome: 853.172 index: 95\n",
      "\n",
      "downloaded genome: 860.11 index: 96\n",
      "\n",
      "downloaded genome: 860.16 index: 97\n",
      "\n",
      "downloaded genome: 860.17 index: 98\n",
      "\n",
      "downloaded genome: 860.18 index: 99\n",
      "\n",
      "downloaded 100 genomes\n"
     ]
    }
   ],
   "source": [
    "all_genomes = bacpacs_sup[GENOME_ID]\n",
    "if download:\n",
    "    download_genomes(all_genomes, bacpacs_patric_files_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "german-territory",
   "metadata": {},
   "source": [
    "# Create fasta file from PATRIC genomes files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "permanent-petite",
   "metadata": {},
   "outputs": [],
   "source": [
    "UNKNOWN_PGFAM = 'X'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "available-audit",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_patric_file(file_path):\n",
    "    \n",
    "    genome_df = pd.read_csv(file_path, sep='\\t')\n",
    "    pgfams = genome_df['pgfam_id'].fillna(UNKNOWN_PGFAM)\n",
    "    \n",
    "    return pgfams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "creative-reference",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_genomes_files(patric_files_folder, output_path):\n",
    "    with os.scandir(patric_files_folder) as entries:\n",
    "        with open(output_path, 'w') as out_f:\n",
    "            for entry in entries:\n",
    "                if entry.is_file():\n",
    "                    genome_id = entry.name.split(FILE_SUFFIX)[0]\n",
    "                    genome_pgfams = parse_patric_file(entry)\n",
    "\n",
    "                    out_f.write(f'>{genome_id}\\n')\n",
    "                    out_f.write('\\n'.join(genome_pgfams))\n",
    "                    out_f.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "smaller-apartment",
   "metadata": {},
   "outputs": [],
   "source": [
    "if write:\n",
    "    write_genomes_files(bacpacs_patric_files_path, bacpacs_genomes_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
